## IMPORT DEPENDENCIES ##

In [35]:
import pandas as pd
import numpy as np
import matplotlib.pyplot  as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


In [3]:
df = pd.read_csv("C:\\Users\\hp\\Downloads\\archive (1)\\creditcard.csv")

In [4]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

## EDA ##

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [6]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [7]:
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

 ### 0 - Normal transaction
 ### 1 - Fraudulent Transaction
### The dataset is  highly imbalanced ###

## Separating the dataset for analysis ##

In [10]:
normal = df[df['Class'] == 0]
fraud = df[df['Class'] == 1]

In [11]:
print(normal.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [12]:
normal.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [13]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [14]:
## Comparing the values for both transactions
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

## Undersampling ##

## Building a dataset containing equal distribution of normal and  fraudulent Transactions ##

In [15]:
## fraud transactions is 492
normal_sample = normal.sample(n=492)

## concatenating the two dataframes ##

In [16]:
new_data = pd.concat([normal_sample, fraud], axis=0)

In [17]:
new_data.head()

Time        V1        V2        V3        V4        V5        V6  \
224694  143921.0  1.882045 -0.499051 -1.494708  0.313994 -0.231542 -1.276645   
139722   83311.0  0.957826 -0.255238  0.131789 -0.004205 -0.234121 -0.395831   
114387   73460.0 -1.478249  0.548563  1.733035  0.605999  0.000109 -0.182800   
269358  163637.0 -0.589759  0.692713  2.038383 -0.549292  0.048791  0.161394   
226555  144685.0 -1.117696  0.509959  0.399289 -0.324950  1.064923  1.327217   

              V7        V8        V9  ...       V21       V22       V23  \
224694  0.275833 -0.371519  0.552662  ...  0.256046  0.568700 -0.056026   
139722  0.271531 -0.078427 -0.434784  ... -0.354404 -1.343374  0.180285   
114387  0.177630  0.579999 -1.123980  ...  0.348411  0.527743  0.014380   
269358  0.409980  0.144724  0.385880  ... -0.198126 -0.338217 -0.348077   
226555  0.186221  1.004275 -0.371318  ...  0.303674  0.668938 -0.063212   

             V24       V25       V26       V27       V28  Amount  Class  
224694 -0.054202  0.004272  0.785494 -0.117107 -0.054626  114.88      0  
139722  0.040108 -0.161975  0.507927 -0.098090  0.019656  135.00      0  
114387  0.131962 -0.132864  0.851805 -0.118350 -0.031664   77.38      0  
269358 -0.490503  0.285887 -0.659436  0.179951 -0.096924    1.00      0  
226555 -1.692618 -0.285458 -0.684666  0.080583  0.041572   49.80      0  

[5 rows x 31 columns]

In [18]:
new_data.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [19]:
new_data['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [20]:
new_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      95532.304878  0.076910  0.046692 -0.054092 -0.098002  0.037905   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.024438  0.028943  0.073415 -0.039635  ... -0.007812 -0.013383   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.042134  0.000617 -0.001684  0.007878  0.008597 -0.012463 -0.002419   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       77.432642  
1      122.211321  

[2 rows x 30 columns]

### shuffling the data ###

In [22]:
new_data = new_data.sample(frac=1,random_state=42).reset_index(drop=True)

In [23]:
X = new_data.drop(columns = 'Class', axis=1)
Y = new_data['Class']

In [24]:
print(X)

         Time         V1        V2         V3        V4        V5        V6  \
0     41233.0 -10.645800  5.918307 -11.671043  8.807369 -7.975501 -3.586806   
1    136607.0   2.063489 -0.161576  -1.937503  0.469826  0.448842 -0.802207   
2     72327.0  -4.198735  0.194121  -3.917586  3.920748 -1.875486 -2.118933   
3    139601.0  -1.021834  1.622710  -0.682330 -1.092109  0.629644 -0.642935   
4     66101.0   0.423354 -1.583684   0.833238  0.488244 -1.589162 -0.042204   
..        ...        ...       ...        ...       ...       ...       ...   
979  144236.0  -1.016855  1.395365  -0.606025 -0.457053  0.162011 -1.296060   
980   81436.0 -11.504569  9.247178  -8.405677 -0.306943 -6.749095 -1.862070   
981  128471.0   0.909124  1.337658  -4.484728  3.245358 -0.417809 -0.762119   
982   66430.0  -1.572200 -0.701301   0.511534 -0.862446  0.263180 -0.262617   
983  158306.0   2.220691 -1.525061  -0.736854 -1.437077 -1.561950 -0.856879   

            V7        V8        V9  ...       V20  

In [25]:
print(Y)

0      1
1      0
2      1
3      0
4      0
      ..
979    0
980    0
981    1
982    0
983    0
Name: Class, Length: 984, dtype: int64


## Splitting the data into training and test data ##

In [26]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y,random_state=3)

In [28]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


## Model Training ##

In [29]:
model = LogisticRegression()

In [32]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [33]:
## training the logistic Regression model with training data ##
model.fit(X_train, Y_train)

C:\Users\hp\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

## Model Evaluation ##

In [36]:
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [37]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9466327827191868


In [38]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [39]:
print('Accuracy on Testing data : ', test_data_accuracy)

Accuracy on Testing data :  0.9441624365482234


## The model performed well with an accuracy score of 94% ##